Port Sexcon from Azure Peak

ComradeNiobe opened the pull request:
<!-- !! PLEASE, READ THIS !! -->
<!-- We recommend to check the contributing page before opening pull requests. -->
<!-- https://github.com/NebulaSS13/Nebula/blob/dev/CONTRIBUTING.md -->
<!-- If you're opening a pull request which changes A LOT of icon/map files: -->
<!-- Add [IDB IGNORE] (to ignore icon file changes) or [MDB IGNORE] (to ignore map file changes) in the PR title. -->
<!-- These tags prevent huge diffs from overloading IconDiffBot and MapDiffBot. -->

## Description of changes
This PR ports Sexcon2 from Azure Peak with heavy adjustments, namely decalization of sex actions. Not everything has been adapted, so still WIP, but it does work!

I also went ahead and stripped out a bunch of other roguetownisms, like sex sounds being emotes, which doesn't really work with Nebula's emote system too well.

## Why and what will this PR improve
Sexcon2 has the most advanced ERP mechanics in SS13, and since our server is ERP allowed, it'd be a nice way to mix it up.

## Authorship
Azure Peak, Ratwood, Azarak.
Comrade Niobe.

## Changelog
:cl:
add: Added sexcon2 content modpack.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

ComradeNiobe committed:
Port Sexcon from Azure

This PR ports Sexcon from Azure Peak with heavy adjustments, namely decalization of sex actions. Not everything has been adapted, so still WIP, but it does work!

noelle-lavenza commented on mods/content/sexcon/_sexcon.dm:
```diff
@@ -0,0 +1,84 @@
+
+/decl/config/toggle/sexcon/Initialize()
+	// Init decls repository for sex actions.
+	decls_repository.get_decls_of_subtype(/decl/sex_action)
```
this is unnecessary, they will get initialised as-needed

noelle-lavenza commented on mods/content/sexcon/datum/sex_action.dm:
```diff
@@ -0,0 +1,44 @@
+	SHOULD_CALL_PARENT(TRUE)
+
+	var/decl/pronouns/pronouns = user.get_pronouns()
+	return pronouns
```
why does this return the user's pronouns? the return value of this proc is never checked

ComradeNiobe replied:
Unfinished implementation at the moment, but the intention is like so:
```dm
/decl/sex_action/vaginal_sex/on_start(mob/living/human/user, mob/living/human/target)
	. = ..()
	var/decl/pronouns/pronouns = .

	user.visible_message(span_warning("[user] slides [pronouns.his] cock into [target]'s cunt!"))
	playsound(target, list('mods/content/sexcon/sounds/mat/insert (1).ogg','mods/content/sexcon/sounds/mat/insert (2).ogg'), 20, TRUE, ignore_walls = FALSE)
```

noelle-lavenza commented on mods/content/sexcon/datum/sexcon.dm:
```diff
@@ -0,0 +1,603 @@
+	desire_stop = FALSE
+	current_action = saction
+	log_attack("[key_name(user)] started sex action on [key_name(target)]: [current_action.name]")
+	INVOKE_ASYNC(src, PROC_REF(sex_action_loop))
```
should probably just add `set waitfor = FALSE` in `sex_action_loop` and then you can call it normally, all INVOKE_ASYNC does is effectively that anyway

ComradeNiobe replied:
Done!

noelle-lavenza commented on mods/content/sexcon/datum/sexcon.dm:
```diff
@@ -0,0 +1,603 @@
+		if(80 to 90)
+			severity = 9
+		if(90 to INFINITY)
+			severity = 10
```
is there a reason this isn't just `severity = clamp(ceil(arousal / 10), 0, 10)`? like I guess if you wanted to adjust it later, but this is clearly just "divide by 10 and round up"

ComradeNiobe replied:
Agreed. Wiped out!

noelle-lavenza commented on mods/content/sexcon/datum/sexcon_helpers.dm:
```diff
@@ -0,0 +1,83 @@
+		vag.be_impregnated(src)
+	*/
+
+/proc/add_cum_floor(turfu)
```
this should be relatively easy to add btw, you can reference vomit code for how to do reagent-based cleanables

noelle-lavenza commented on mods/content/sexcon/datum/sex_action.dm:
```diff
@@ -0,0 +1,44 @@
+	SHOULD_CALL_PARENT(TRUE)
+
+	var/decl/pronouns/pronouns = user.get_pronouns()
+	return pronouns
```
why does this return the user's pronouns? the return value of this proc is never checked

ComradeNiobe replied:
Ditto as below

noelle-lavenza commented on mods/content/sexcon/datum/sex_actions/deviant/footjob.dm:
```diff
@@ -0,0 +1,42 @@
+/decl/sex_action/footjob/can_perform(mob/living/human/user, mob/living/human/target)
+	if(user == target)
+		return FALSE
+	if(!get_location_accessible(user, BODY_ZONE_PRECISE_L_FOOT))
+		return FALSE
+	if(!get_location_accessible(user, BODY_ZONE_PRECISE_R_FOOT))
+		return FALSE
+	if(!get_location_accessible(target, BODY_ZONE_PRECISE_GROIN, TRUE))
+		return FALSE
+	if(!target.getorganslot(ORGAN_SLOT_PENIS))
+		return FALSE
```
i'm not gonna go through for each one, but here's how I'd do it The Nebula Way™ and this applies to other ones. note that `BP_PENIS` would still need to be defined somewhere, and the body part would need to be added
```suggestion
	var/user_covered_parts = user.get_covered_body_parts()
	if(user_covered_parts & (SLOT_FOOT_LEFT|SLOT_FOOT_RIGHT)) // fail if either are covered
		return FALSE
	var/target_covered_parts = target.get_covered_body_parts()
	if(user_covered_parts & SLOT_LOWER_BODY)
		return FALSE
	if(!target.has_organ(BP_PENIS))
		return FALSE
```

ComradeNiobe replied:
This will make things much easier for me indeed!

noelle-lavenza commented on mods/content/sexcon/datum/sex_actions/deviant/facesitting.dm:
```diff
@@ -0,0 +1,45 @@
+		return FALSE
+	// Target can't stand up
+	if(!target.resting)
+		return FALSE
```
```suggestion
	// Need to be standing
	if(user.current_posture.prone)
		return FALSE
	// Target can't be standing
	if(!target.current_posture.prone)
		return FALSE
```
`resting` is `current_posture.prone` on Nebula btw, to support things like drakes being able to lay, sit, and stand

noelle-lavenza commented on mods/content/sexcon/_sexcon.dm:
```diff
@@ -0,0 +1,84 @@
+
+#define SEX_ACTION(sex_action_type) RESOLVE_TO_DECL(sex_action_type)
+
+#define MAX_AROUSAL 150
```
these are limited to just the modpack so it's *probably* fine, but i would personally prefer `const` vars on `/datum/sex_controller` or wherever else it's used, assuming it's only used on one type (if it's only used in one proc, it can be a proc-scope const even!)

noelle-lavenza commented on mods/content/sexcon/datum/sex_action.dm:
```diff
@@ -0,0 +1,44 @@
+	/// Whether the action requires an aggressive grab on the victim
+	var/require_grab = FALSE
+	/// If a grab is required, this is the required state of it
+	var/required_grab_state = /decl/grab/normal/aggressive // ? This is obsoleted by the danger_grab flag.
```
i'm not sure if it is, `force_danger` on a grab is specifically for things like slamming someone's head into a table. then again it does kind of represent "you can force someone to do something they're actively resisting" so i think you're right actually

noelle-lavenza commented on mods/content/sexcon/datum/sexcon.dm:
```diff
@@ -0,0 +1,603 @@
+	switch(href_list["task"])
+		if("action")
+			var/action_path = href_list["action_type"]
+			var/decl/sex_action/saction = decls_repository.get_decl_by_id(action_path)
```
fwiw this will return an instance and not a type, whereas other code seems to assume `try_start_action()` receives a decl instance. it's probably easier to just make *everything* use decl instances

noelle-lavenza commented on mods/content/sexcon/datum/sexcon_helpers.dm:
```diff
@@ -0,0 +1,83 @@
+	if(!istype(user) || !istype(target))
+		return
+	if(!user.can_do_sex())
+		to_chat(user, SPAN_WARNING("I can't do this.</span>"))
```
```suggestion
		to_chat(user, SPAN_WARNING("I can't do this."))
```

noelle-lavenza commented on mods/content/sexcon/datum/sexcon.dm:
```diff
@@ -0,0 +1,603 @@
+			return pick(list("brutally", "violently", "relentlessly", "savagely", "mercilessly"))
+
+/datum/sex_controller/proc/spanify_force(string)
+	switch(force)
+		if(SEX_FORCE_LOW)
+			return "<span class='love_low'>[string]</span>"
+		if(SEX_FORCE_MID)
+			return "<span class='love_mid'>[string]</span>"
+		if(SEX_FORCE_HIGH)
+			return "<span class='love_high'>[string]</span>"
+		if(SEX_FORCE_EXTREME)
+			return "<span class='love_extreme'>[string]</span>"
```
```suggestion
	var/force_span
	switch(force)
		if(SEX_FORCE_LOW)
			force_span = "love_low"
		if(SEX_FORCE_MID)
			force_span = "love_mid"
		if(SEX_FORCE_HIGH)
			force_span = "love_high"
		if(SEX_FORCE_EXTREME)
			force_span = "love_extreme"
	return SPAN_CLASS(force_span, string)
```
personally i'd prefer to do it this way, although if you add `SPAN_LOVE_LOW` defines those might be more preferable

noelle-lavenza commented on mods/content/sexcon/datum/sexcon_helpers.dm:
```diff
@@ -0,0 +1,83 @@
+	. = ..()
+	QDEL_NULL(sexcon)
+
+/mob/living/human/verb/start_sexcon()
```
i recommend pulling the insides of this proc into a proc like `start_sexcon(mob/living/human/target)` and renaming this to like, `start_sexcon_verb()`. then you can do something like
```dm
/mob/living/human/handle_mouse_drop(atom/over, mob/user)
	. = ..()
	if(!. && ishuman(over) && ishuman(user) && user == src && user != over)
		start_sexcon(over)
		return TRUE
```
to give the middle-mouse-drag interaction that other codebases have

noelle-lavenza commented on mods/content/sexcon/datum/sexcon_helpers.dm:
```diff
@@ -0,0 +1,83 @@
+	if(!user.can_do_sex())
+		to_chat(user, SPAN_WARNING("I can't do this.</span>"))
+		return
+	if(!target.client?.prefs || (target.get_preference_value(/datum/client_preference/sexable) == PREF_NO)) // Don't bang someone that dosn't want it.
```
this is a little simpler; if they don't have a client or don't have prefs it'll return null, which isn't `PREF_YES`. checking `!= PREF_YES` is more exhaustive, we only want to proceed if their prefs explicitly say yes
(although this WILL make it so that clientless mobs use the default pref. i would probably check for client explicitly elsewhere)
```suggestion
	if(target.get_preference_value(/datum/client_preference/sexable) != PREF_YES)) // Don't bang someone that doesn't want it.
```

ComradeNiobe replied:
Applied manually

noelle-lavenza commented on mods/content/sexcon/_sexcon.dm:
```diff
@@ -0,0 +1,84 @@
+/datum/client_preference/sexable
+	description = "Sex Mechanics"
+	key = "SEXCON_TOGGLE"
+	options = list(PREF_NO, PREF_YES)
```
```suggestion
	default_value = PREF_NO
```
you could also just use the default order but have it default to `PREF_NO`. these are identical except one swaps the display order.

noelle-lavenza commented on mods/content/sexcon/datum/sexcon.dm:
```diff
@@ -0,0 +1,603 @@
+
+/datum/sex_controller/proc/sex_action_loop()
+	// Do action loop
+	var/decl/sex_action/performed_action_type = current_action
```
as i mentioned elsewhere `current_action` is always set to an instance rather than a type, so this will not actually be a type.

ComradeNiobe replied:
Cut it out entirely.

noelle-lavenza reviewed (changes requested):
some preliminary review, there's still a bunch of roguetownisms that need to be gone through but I assume you're working on those already

ComradeNiobe committed:
Remove redundant decl init

ComradeNiobe committed:
Several adjustments

Simplifies update_pink_screen and makes the saction loop call directly with WAITFOR false.

ComradeNiobe committed:
Reduce redundant proc use, refactor more checks.

ComradeNiobe committed:
Update mods/content/sexcon/datum/sexcon_helpers.dm

Co-authored-by: Sutures <101909198+noelle-lavenza@users.noreply.github.com>

ComradeNiobe committed:
Update mods/content/sexcon/_sexcon.dm

Co-authored-by: Sutures <101909198+noelle-lavenza@users.noreply.github.com>

noelle-lavenza commented on mods/content/sexcon/datum/sexcon.dm:
```diff
@@ -0,0 +1,581 @@
+		var/grab_bypass = (saction.aggro_grab_instead_same_tile && target.has_danger_grab(user))
+		if(!same_tile && !grab_bypass)
+			return FALSE
+	if(saction.require_grab)
```
this makes it so that all actions that require grabs require aggro grabs, which i don't think is necessarily intended? you might want one to require just any sort of grab. like. idk. make groping require grabbing someone first or something

ComradeNiobe replied:
The reason behind this is that Roguetown's implementation, `required_grab_state`, isn't actually used in any sex actions at all and stays permanently defaulted to GRAB_AGGRESSIVE and above. I wanted to keep parity in behavior so I made it use `has_danger_grab` instead which achieves the same effect.

If we want to make certain sex actions like groping more interactive though, would certainly be the way to go.

ComradeNiobe committed:
make tests actually run

ComradeNiobe committed:
Update mods/content/sexcon/datum/sexcon.dm

Co-authored-by: Sutures <101909198+noelle-lavenza@users.noreply.github.com>

ComradeNiobe committed:
How did I miss this?

ComradeNiobe committed:
Apply suggestions from code review

Co-authored-by: Sutures <101909198+noelle-lavenza@users.noreply.github.com>

ComradeNiobe committed:
Sexcon3 - NEBULAIZED EDITION

ComradeNiobe committed:
fix include order

ComradeNiobe committed:
whoops,, that'd be awkward...
